### PR TITLE
MM-21157: Triggers role sync when a syncable is linked, unlinked, or updated.

### PR DESCRIPTION
**Pull request by mkraft**
#### Ticket Link

https://mattermost.atlassian.net/browse/MM-21157

#### Dependency

https://github.com/mattermost/enterprise/pull/555

### CODE REVIEW

File: store/timer_layer.go [hunk 1]
```diff
@@ -568,6 +568,22 @@ func (s *TimerLayerChannelStore) ClearCaches() {
 	return
 }
```
Comment (mkraft) [inline store/timer_layer.go]: My execution of `make store-layers` seems to have picked up some unrelated forgotten new methods.

File: store/sqlstore/group_store.go [hunk 2]
```diff
@@ -851,7 +863,7 @@ func (s *SqlGroupStore) ChannelMembersToRemove() ([]*model.ChannelMember, *model
 					AND GroupMembers.DeleteAt = 0
 				GROUP BY
 					Channels.Id,
-					GroupMembers.UserId)`
+					GroupMembers.UserId)`, channelScope)
```
Comment (mkraft) [inline store/sqlstore/group_store.go]: I’m wondering if the newly added branching statements in this file merit converting the SQL statements to Squirrel.

File: app/syncables.go [hunk 3]
```diff
@@ -101,22 +172,65 @@ func (a *App) DeleteGroupConstrainedMemberships() error {
...
-		a.Log.Info("removed teammember",
-			mlog.String("user_id", userTeam.UserId),
-			mlog.String("team_id", userTeam.TeamId),
-		)
+	err = updateFunc(syncableID, permittedAdmins, store.Equals, true)
+	if err != nil {
+		return err
+	}
+
+	err = updateFunc(syncableID, permittedAdmins, store.NotEquals, false)
```
Comment (sbishel) [inline app/syncables.go]: I'm wondering if "UpdateMembersRole" should run both queries, rather than calling twice.  Is there ever a time you would only set one set and not the other?

File: store/sqlstore/group_store.go (hunk 2 above)
Comment (mkraft) [reply store/sqlstore/group_store.go]: Ok, I'm going to change it.

File: app/syncables.go (hunk 3 above)
Comment (mkraft) [reply app/syncables.go]: My original thinking was to keep the data store "dumber" but now that you've question it I'm changing my mind... it could be a single SQL `UPDATE` where everything `IN` has a boolean value and everything `NOT IN` has the opposite value. I'll make this change. Thanks.

Comment (mkraft) [reply app/syncables.go]: I have [changed the `UpdateMembersRole`](https://github.com/mattermost/mattermost-server/pull/13361/commits/8670e7815d6dc13639b08350f603fec9c1218ab1) methods in a separate PR that this is branched off of. It's using a SQL case statement to conditionally update the `SchemeAdmin` column value in a single query.

**Comment by mkraft** [comment]
```diff
diff --git a/store/sqlstore/group_store.go b/store/sqlstore/group_store.go
index f34d2dc84..1b9e304e3 100644
--- a/store/sqlstore/group_store.go
+++ b/store/sqlstore/group_store.go
@@ -724,53 +724,25 @@ func groupSyncableToGroupChannel(groupSyncable *model.GroupSyncable) *groupChann
 }
 
 func (s *SqlGroupStore) TeamMembersToRemove(teamID *string) ([]*model.TeamMember, *model.AppError) {
-	var teamScope string
+	query := s.getQueryBuilder().Select("TeamMembers.TeamId, TeamMembers.UserId, TeamMembers.Roles, TeamMembers.DeleteAt, TeamMembers.SchemeUser, TeamMembers.SchemeAdmin, (TeamMembers.SchemeGuest IS NOT NULL AND TeamMembers.SchemeGuest) as SchemeGuest").
+		From("TeamMembers").
+		Join("Teams ON Teams.Id = TeamMembers.TeamId").
+		LeftJoin("Bots ON Bots.UserId = TeamMembers.UserId").
+		Where(sq.Eq{"TeamMembers.DeleteAt": 0, "Teams.DeleteAt": 0, "Teams.GroupConstrained": true, "Bots.UserId": nil}).
+		Where("(TeamMembers.TeamId, TeamMembers.UserId) NOT IN (SELECT Teams.Id AS TeamId, GroupMembers.UserId FROM Teams JOIN GroupTeams ON GroupTeams.TeamId = Teams.Id JOIN UserGroups ON UserGroups.Id = GroupTeams.GroupId JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id WHERE Teams.GroupConstrained = TRUE AND GroupTeams.DeleteAt = 0 AND UserGroups.DeleteAt = 0 AND Teams.DeleteAt = 0 AND GroupMembers.DeleteAt = 0 GROUP BY Teams.Id, GroupMembers.UserId)")
 
 	if teamID != nil {
-		teamScope = fmt.Sprintf("AND TeamMembers.TeamId = '%s'", *teamID)
+		query = query.Where(sq.Eq{"TeamMembers.TeamId": *teamID})
 	}
 
-	sql := fmt.Sprintf(`
-		SELECT
-			TeamMembers.TeamId,
-			TeamMembers.UserId,
-			TeamMembers.Roles,
-			TeamMembers.DeleteAt,
-			TeamMembers.SchemeUser,
-			TeamMembers.SchemeAdmin,
-			(TeamMembers.SchemeGuest IS NOT NULL AND TeamMembers.SchemeGuest) as SchemeGuest
-		FROM
-			TeamMembers
-			JOIN Teams ON Teams.Id = TeamMembers.TeamId
-			LEFT JOIN Bots ON Bots.UserId = TeamMembers.UserId
-		WHERE
-			TeamMembers.DeleteAt = 0
-			AND Teams.DeleteAt = 0
-			AND Teams.GroupConstrained = TRUE
-			AND Bots.UserId IS NULL
-			%s
-			AND (TeamMembers.TeamId, TeamMembers.UserId)
-			NOT IN (
-				SELECT
-					Teams.Id AS TeamId, GroupMembers.UserId
-				FROM
-					Teams
-					JOIN GroupTeams ON GroupTeams.TeamId = Teams.Id
-					JOIN UserGroups ON UserGroups.Id = GroupTeams.GroupId
-					JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id
-				WHERE
-					Teams.GroupConstrained = TRUE
-					AND GroupTeams.DeleteAt = 0
-					AND UserGroups.DeleteAt = 0
-					AND Teams.DeleteAt = 0
-					AND GroupMembers.DeleteAt = 0
-				GROUP BY
-					Teams.Id,
-					GroupMembers.UserId)`, teamScope)
+	sql, params, err := query.ToSql()
+	if err != nil {
+		return nil, model.NewAppError("SqlGroupStore.TeamMembersToRemove", "store.sql_group.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
 
 	var teamMembers []*model.TeamMember
 
-	_, err := s.GetReplica().Select(&teamMembers, sql)
+	_, err = s.GetReplica().Select(&teamMembers, sql, params...)
 	if err != nil {
 		return nil, model.NewAppError("SqlGroupStore.TeamMembersToRemove", "store.select_error", nil, "", http.StatusInternalServerError)
 	}
@@ -818,56 +790,25 @@ func (s *SqlGroupStore) GetGroupsByChannel(channelId string, opts model.GroupSea
 }
 
 func (s *SqlGroupStore) ChannelMembersToRemove(channelID *string) ([]*model.ChannelMember, *model.AppError) {
-	var channelScope string
+	query := s.getQueryBuilder().Select("ChannelMembers.ChannelId, ChannelMembers.UserId, ChannelMembers.LastViewedAt, ChannelMembers.MsgCount, ChannelMembers.MentionCount, ChannelMembers.NotifyProps, ChannelMembers.LastUpdateAt, ChannelMembers.LastUpdateAt, ChannelMembers.SchemeUser, ChannelMembers.SchemeAdmin, (ChannelMembers.SchemeGuest IS NOT NULL AND ChannelMembers.SchemeGuest) as SchemeGuest").
+		From("ChannelMembers").
+		Join("Channels ON Channels.Id = ChannelMembers.ChannelId").
+		LeftJoin("Bots ON Bots.UserId = ChannelMembers.UserId").
+		Where(sq.Eq{"Channels.DeleteAt": 0, "Channels.GroupConstrained": true, "Bots.UserId": nil}).
+		Where("(ChannelMembers.ChannelId, ChannelMembers.UserId) NOT IN (SELECT Channels.Id AS ChannelId, GroupMembers.UserId FROM Channels JOIN GroupChannels ON GroupChannels.ChannelId = Channels.Id JOIN UserGroups ON UserGroups.Id = GroupChannels.GroupId JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id WHERE Channels.GroupConstrained = TRUE AND GroupChannels.DeleteAt = 0 AND UserGroups.DeleteAt = 0 AND Channels.DeleteAt = 0 AND GroupMembers.DeleteAt = 0 GROUP BY Channels.Id, GroupMembers.UserId)")
 
 	if channelID != nil {
-		channelScope = fmt.Sprintf("AND ChannelMembers.ChannelId = '%s'", *channelID)
+		query = query.Where(sq.Eq{"ChannelMembers.ChannelId": *channelID})
 	}
 
-	sql := fmt.Sprintf(`
-		SELECT
-			ChannelMembers.ChannelId,
-			ChannelMembers.UserId,
-			ChannelMembers.LastViewedAt,
-			ChannelMembers.MsgCount,
-			ChannelMembers.MentionCount,
-			ChannelMembers.NotifyProps,
-			ChannelMembers.LastUpdateAt,
-			ChannelMembers.LastUpdateAt,
-			ChannelMembers.SchemeUser,
-			ChannelMembers.SchemeAdmin,
-			(ChannelMembers.SchemeGuest IS NOT NULL AND ChannelMembers.SchemeGuest) as SchemeGuest
-		FROM
-			ChannelMembers
-			JOIN Channels ON Channels.Id = ChannelMembers.ChannelId
-			LEFT JOIN Bots ON Bots.UserId = ChannelMembers.UserId
-		WHERE
-			Channels.DeleteAt = 0
-			AND Channels.GroupConstrained = TRUE
-			AND Bots.UserId IS NULL
-			%s
-			AND (ChannelMembers.ChannelId, ChannelMembers.UserId)
-			NOT IN (
-				SELECT
-					Channels.Id AS ChannelId, GroupMembers.UserId
-				FROM
-					Channels
-					JOIN GroupChannels ON GroupChannels.ChannelId = Channels.Id
-					JOIN UserGroups ON UserGroups.Id = GroupChannels.GroupId
-					JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id
-				WHERE
-					Channels.GroupConstrained = TRUE
-					AND GroupChannels.DeleteAt = 0
-					AND UserGroups.DeleteAt = 0
-					AND Channels.DeleteAt = 0
-					AND GroupMembers.DeleteAt = 0
-				GROUP BY
-					Channels.Id,
-					GroupMembers.UserId)`, channelScope)
+	sql, params, err := query.ToSql()
+	if err != nil {
+		return nil, model.NewAppError("SqlGroupStore.ChannelMembersToRemove", "store.sql_group.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
 
 	var channelMembers []*model.ChannelMember
 
-	_, err := s.GetReplica().Select(&channelMembers, sql)
+	_, err = s.GetReplica().Select(&channelMembers, sql, params...)
 	if err != nil {
 		return nil, model.NewAppError("SqlGroupStore.ChannelMembersToRemove", "store.select_error", nil, err.Error(), http.StatusInternalServerError)
 	}
```

File: store/sqlstore/group_store.go (hunk 2 above)
Comment (mkraft) [reply store/sqlstore/group_store.go]: Done.

Comment (catalintomai) [reply store/sqlstore/group_store.go]: For me, at least in this case, I find the original SQL statements approach (much) more readable (you can easily read the SQL code at once, and add comments if needed) when compared to Squirrel.

Comment (mkraft) [reply store/sqlstore/group_store.go]: > For me, at least in this case, I find the original SQL statements approach (much) more readable (you can easily read the SQL code at once, and add comments if needed) when compared to Squirrel.

Couldn't agree more. It was harder to write and is harder to read.

**Comment by mkraft** [comment]
This PR needs to incorporate the cache clearing logic as done in https://github.com/mattermost/enterprise/pull/548/commits/44f8ab0fc9667458b919c2b657361f9ef35c1614.

**Comment by srkgupta** [comment]
Hi @mkraft 
As you had mentioned, this PR needs the logic to clear cache whenever the changes are made. Can you please incorporate the same and let me know so that this could all be tested and PR merged.

File: store/storetest/mocks/ChannelStore.go [hunk 4]
```diff
@@ -1173,16 +1173,6 @@ func (_m *ChannelStore) InvalidateMemberCount(channelId string) {
...
-// InvalidateMembersForAllUsers provides a mock function with given fields:
-func (_m *ChannelStore) InvalidateMembersForAllUsers() {
-	_m.Called()
-}
-
-// InvalidateMembersForUser provides a mock function with given fields: userId
-func (_m *ChannelStore) InvalidateMembersForUser(userId string) {
-	_m.Called(userId)
-}
-
```
Comment (mkraft) [inline store/storetest/mocks/ChannelStore.go]: Not part of this PR, but these were just missing.

**Comment by mkraft** [comment]
> Hi @mkraft
> As you had mentioned, this PR needs the logic to clear cache whenever the changes are made. Can you please incorporate the same and let me know so that this could all be tested and PR merged.

I have added the cached-busting code. It's ready for you.

File: api4/group.go [hunk 5]
```diff
@@ -192,6 +192,10 @@ func linkGroupSyncable(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Not awaiting completion because the group sync job executes the same procedure—but for all syncables—and
+	// persists the execution status to the jobs table.
+	go c.App.SyncRolesAndMembership(syncableID, syncableType)
```
Comment (mkraft) [inline api4/group.go]: If these API calls are invoked via the CLI these calls to `go c.App.SyncRolesAndMembership` may not complete.

**Comment by mkraft** [comment]
> Hi @mkraft
> Currently I am seeing cache issue when a role is updated for a Channel.
> 
> Steps to reproduce:
> a) On a browser, login as a LDAP user, say User1, who belongs to Developers Group.
> b) On another browser, login as another LDAP user, say User2 who belongs to Developers & Security Engineers Group.
> c) For a Group Synced Channel, first assign Channel Admin to Channel1 for both roles Developers & Security.
> d) Check that user is assigned to Channel1 as expected.
> e) Now try to modify the role of Developers to Channel Member and save.
> f) Note that User1's permission was not updated. User1 is still shown as Channel Admin instead of Channel Member.
> 
> This works fine in case of Team Role Update. However for Channel, it is not working. Can you please check

This is an existing production bug unrelated to this feature and can be reproduced without groups. 

Steps:

1. Login as system admin.
2. Create a channel. 
3. Create two other users (user1 and user2) and invite them both to the channel.
4. Login as user1 in another browser "Manage Members" from the channel menu.
5. As system admin also select "Manage Members" from the channel dropdown.
6. As system admin update user2's role.

Actual: user1 does not see the role change for user2 in the members modal. A page refresh shows the correct state that matches the database.

**Comment by srkgupta** [comment]
Thanks @mkraft. The issue has been raised separately and is tracked here:
https://mattermost.atlassian.net/browse/MM-21517
